lib/wasix/src/syscalls/wasix/epoll_ctl.rs

```diff
@@ -0,0 +1,179 @@
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc::UnboundedSender;
+use wasmer_wasix_types::wasi::{
+    EpollCtl, EpollEvent, EpollType, SubscriptionClock, SubscriptionUnion, Userdata,
+};
+
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+
+use futures::Future;
+
+use super::*;
+use crate::{
+    fs::{EpollFd, InodeValFilePollGuard, InodeValFilePollGuardJoin},
+    state::PollEventSet,
+    syscalls::*,
+    WasiInodes,
+};
+
+/// ### `epoll_ctl()`
+/// Modifies an epoll interest list
+/// Output:
+/// - `Fd fd`
+///   The new file handle that is used to modify or wait on the interest list
+#[instrument(level = "trace", skip_all, fields(timeout_ms = field::Empty, fd_guards = field::Empty, seen = field::Empty, fd), ret, err)]
+pub fn epoll_ctl<M: MemorySize + 'static>(
+    mut ctx: FunctionEnvMut<'_, WasiEnv>,
+    epfd: WasiFd,
+    op: EpollCtl,
+    fd: WasiFd,
+    event_ref: WasmPtr<EpollEvent<M>, M>,
+) -> Result<Errno, WasiError> {
+    let env = ctx.data();
+
+    let memory = unsafe { env.memory_view(&ctx) };
+    let event = wasi_try_mem_ok!(event_ref.read(&memory));
+
+    let fd_entry = wasi_try_ok!(env.state.fs.get_fd(epfd));
+
+    let inode = fd_entry.inode.clone();
+    let tasks = env.tasks().clone();
+    let mut inode_guard = inode.read();
+    match inode_guard.deref() {
+        Kind::Epoll {
+            subscriptions, tx, ..
+        } => {
+            if let EpollCtl::Del | EpollCtl::Mod = op {
```

Comment:
As mentioned before, can we collect all the related business logic in a `EpollObject` or similar struct?

Reply:
that is better done in the fs refactor, I'll just keeping it within the existing technical debt